ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java

```diff
@@ -1496,7 +1499,13 @@ public static void mvFileToFinalPath(Path specPath, Configuration hconf,
           // for CTAS or Create MV statements
           perfLogger.perfLogBegin("FileSinkOperator", "moveSpecifiedFileStatus");
           LOG.debug("CTAS/Create MV: Files being renamed:  " + filesKept.toString());
-          moveSpecifiedFilesInParallel(hconf, fs, tmpPath, specPath, filesKept);
+          if (conf.getTable() != null && conf.getTable().getTableType().equals(TableType.EXTERNAL_TABLE)) {
+            // Do this optimisation only for External tables.
+            createFileList(filesKept, tmpPath, specPath, fs);
+          } else {
+            List<String> filesKeptPaths = filesKept.stream().map(x -> x.getPath().toString()).collect(Collectors.toList());
```

Comment:
filesKept is a set. Is it possible to retain as "set" after the mapping? (Something like Collectors.toSet()). if so, no need to change the signature of moveSpecifiedFilesInParallel?

Reply:
Done, Changed to Set.